### PR TITLE
Adds userspace, updates planck to use it and adds crkbd support for snowkuma

### PR DESCRIPTION
**Pull request by snowkuma**
## Description

Adds userspace to allow me to easily setup multiple keyboards with the same base layouts.
Additional tidying and reorganisation of my previous planck layout.
Adds crkbd layout for my userid.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: users/snowkuma/snowkuma.c [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+#endif
+void keyboard_post_init_user() {
+  #ifdef RGBLIGHT_ENABLE
+     // Cycles through the entire hue wheel and resetting to default color
+     uint16_t default_hue = rgblight_config.hue;
+     rgblight_enable_noeeprom(); 
+     layer_state_set_user(layer_state);
+     rgblight_mode_noeeprom(RGBLIGHT_MODE_STATIC_LIGHT);
+     for (uint16_t i = 255; i > 0; i--) {
+          rgblight_sethsv_noeeprom((i + default_hue) % 255, rgblight_config.sat, rgblight_config.val);
```
Comment (drashna) [inline users/snowkuma/snowkuma.c:35]:
```suggestion
          rgblight_sethsv_noeeprom((i + default_hue) % 255, rgblight_get_sat(), rgblight_get_val());
```

File: users/snowkuma/tap_dances.c [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+};
+
+void zoom_toggle(qk_tap_dance_state_t *state, void *user_data) {
+  if (state->count == 1) {
+    register_code(KC_LGUI);
+    register_code(KC_LSFT);
+    register_code(KC_A);
+    unregister_code(KC_A);
+    unregister_code(KC_LSFT);
+    unregister_code(KC_LGUI);
```
Comment (drashna) [inline users/snowkuma/tap_dances.c:23]: Can replace all of this with just: 
```suggestion
    tap_code16(G(S(KC_A)));
```

**Comment by drashna** [comment]
Also, I think this should be targtting `master`, rather than `develop`.

File: users/snowkuma/snowkuma.c [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+#include "snowkuma.h"
+
+layer_state_t layer_state_set_user (layer_state_t state) {
+  state = update_tri_layer_state(state, _NAV, _SYMBOL, _MOUSE);
+  state = update_tri_layer_state(state, _NUMBER, _THUMB, _FUNCTION);
+  return state;
+}
+
+#ifdef RGBLIGHT_ENABLE
+extern rgblight_config_t rgblight_config;
```
[the remark this answers is not in the document]
Comment (snowkuma) [reply users/snowkuma/snowkuma.c:25]: 👍

File: users/snowkuma/snowkuma.c [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+#endif
+void keyboard_post_init_user() {
+  #ifdef RGBLIGHT_ENABLE
+     // Cycles through the entire hue wheel and resetting to default color
+     uint16_t default_hue = rgblight_config.hue;
```
[the remark this answers is not in the document]
Comment (snowkuma) [reply users/snowkuma/snowkuma.c:30]: 👍

File: users/snowkuma/snowkuma.c (hunk 1 above)
Comment (snowkuma) [reply users/snowkuma/snowkuma.c:35]: 👍

File: users/snowkuma/tap_dances.c (hunk 2 above)
Comment (snowkuma) [reply users/snowkuma/tap_dances.c:23]: 👍 That's cool I hadn't realised we could do that 😀

**Comment by snowkuma** [comment]
Thanks for taking the time to look at this PR 🙏
Do you think I should raise a new PR targeting `master` instead?

**Comment by drashna** [comment]
No need.  you can edit the title, and retarget the PR.  Well, I'm not sure if you can, but I have done so, so should be good.

**Comment by drashna** [comment]
Also, you may want to run a clang-format pass of your files. There are a lot of mixing of tap and space, and inconsistent spacing. 

You can run `qmk cformat` on the files, and this will handle things for you (if you're not sure how to).

**Comment by snowkuma** [comment]
one of the cformat runs inadvertantly changed some of the core files so I reverted that commit.

File: users/snowkuma/rules.mk [hunk 5]
```diff
@@ -0,0 +1,13 @@
+BOOTMAGIC_ENABLE = no       # Disable Boot Magic (https://beta.docs.qmk.fm/features/feature_bootmagic)
+MOUSEKEY_ENABLE = yes       # Mouse keys(+4700)
+EXTRAKEY_ENABLE = yes       # Audio control and System control(+450)
+TAP_DANCE_ENABLE = yes      # Enable Tap Dance.
```
Comment (drashna) [inline users/snowkuma/rules.mk:4]:
```suggestion
BOOTMAGIC_ENABLE = no       # Disable Boot Magic (https://docs.qmk.fm/#/feature_bootmagic)
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
TAP_DANCE_ENABLE = yes      # Enable Tap Dance.
```

File: users/snowkuma/snowkuma.c [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    layer_state_set_user(layer_state);
+    rgblight_mode_noeeprom(RGBLIGHT_MODE_STATIC_LIGHT);
+    for (uint16_t i = 255; i > 0; i--) {
+        rgblight_sethsv_noeeprom((i + default_hue) % 255, rgblight_get_sat(), rgblight_get_val());
+        matrix_scan();
+        wait_ms(10);
+    }
+#endif
+    layer_state_set_user(layer_state);
+}
```
Comment (drashna) [inline users/snowkuma/snowkuma.c:38]: Just a heads up, I have a better way to do this. It's more complicated, but ... it doesn't block the keyboard from starting or running tasks.
https://github.com/drashna/qmk_firmware/blob/1d4bfbc6d3d5b77a8a8cd3c847b4ea095b2163c4/users/drashna/rgb_stuff.c#L104-L151

Comment (snowkuma) [reply users/snowkuma/snowkuma.c:38]: Nice, I'll take a look 😀

Comment (drashna) [reply users/snowkuma/snowkuma.c:38]: Have you had a chance to take a look at the code I linked? 
It works much better, and will work with split keyboards (without any weird issues).

File: users/snowkuma/tap_dances.c [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    if (state->count == 1) {
+        SEND_STRING(SS_LGUI(" "));
+        SEND_STRING("TypeFu.app\n");
+    } else if (state->count > 1) {
+        SEND_STRING(SS_LGUI(" "));
+        SEND_STRING("Telegram.app\n");
+    }
+}
+
+qk_tap_dance_action_t tap_dance_actions[] = {[TD_ANKASS] = ACTION_TAP_DANCE_FN(anki_assimil), [TD_COPYCUT] = ACTION_TAP_DANCE_FN(copy_cut), [TD_LBRC_BACK] = ACTION_TAP_DANCE_DOUBLE(KC_LBRC, LGUI(KC_LBRC)), [TD_RBRC_FWD] = ACTION_TAP_DANCE_DOUBLE(KC_RBRC, LGUI(KC_RBRC)), [TD_KRAKEN] = ACTION_TAP_DANCE_FN(kraken_safe_reset), [TD_TILD] = ACTION_TAP_DANCE_FN(tilde_home), [TD_TYPTEL] = ACTION_TAP_DANCE_FN(typefu_telegram), [TD_ZOOM] = ACTION_TAP_DANCE_FN(zoom_toggle)};
```
Comment (drashna) [inline users/snowkuma/tap_dances.c:62]: you may want to format this so it's easier to read: 
```suggestion
qk_tap_dance_action_t tap_dance_actions[] = {
    [TD_ANKASS] = ACTION_TAP_DANCE_FN(anki_assimil), 
    [TD_COPYCUT] = ACTION_TAP_DANCE_FN(copy_cut), 
    [TD_LBRC_BACK] = ACTION_TAP_DANCE_DOUBLE(KC_LBRC, LGUI(KC_LBRC)), 
    [TD_RBRC_FWD] = ACTION_TAP_DANCE_DOUBLE(KC_RBRC, LGUI(KC_RBRC)), 
    [TD_KRAKEN] = ACTION_TAP_DANCE_FN(kraken_safe_reset), 
    [TD_TILD] = ACTION_TAP_DANCE_FN(tilde_home), 
    [TD_TYPTEL] = ACTION_TAP_DANCE_FN(typefu_telegram), 
    [TD_ZOOM] = ACTION_TAP_DANCE_FN(zoom_toggle)
};
```

File: users/snowkuma/tap_dances.h [hunk 8]
```diff
@@ -0,0 +1,14 @@
+#pragma once
+#include "snowkuma.h"
+
+// Tap Dance Declarations
+enum { TD_ANKASS = 0, TD_COPYCUT, TD_KRAKEN, TD_LBRC_BACK, TD_RBRC_FWD, TD_TILD, TD_TYPTEL, TD_ZOOM };
```
Comment (drashna) [inline users/snowkuma/tap_dances.h:5]:
```suggestion
enum { 
    TD_ANKASS, 
    TD_COPYCUT, 
    TD_KRAKEN, 
    TD_LBRC_BACK, 
    TD_RBRC_FWD, 
    TD_TILD, 
    TD_TYPTEL, 
    TD_ZOOM, 
};
```

File: users/snowkuma/tap_dances.c (hunk 7 above)
Comment (snowkuma) [reply users/snowkuma/tap_dances.c:62]: Thanks for spotting these, I think it was a result of running CFormat.  I'll check through other files now to make sure nothing else was reduced to monster lines!

File: users/snowkuma/snowkuma.c (hunk 6 above)
Comment (snowkuma) [reply users/snowkuma/snowkuma.c]: Thanks have added a lot of your rgb_stuff to the project.  Awesome stuff!

File: users/snowkuma/tap_dances.c (hunk 7 above)
Comment (drashna) [reply users/snowkuma/tap_dances.c]: You can use comments to disable the formatting. 

```c
// clang-format off 

// clang-format on
```

File: keyboards/crkbd/keymaps/snowkuma/config.h [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#define TAPPING_TERM 200
+
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/snowkuma/config.h:24]:
```suggestion
```

File: keyboards/crkbd/keymaps/snowkuma/rules.mk [hunk 10]
```diff
@@ -0,0 +1,7 @@
+RGB_MATRIX_ENABLE = WS2812
+MOUSEKEY_ENABLE = no
+OLED_DRIVER_ENABLE = yes
+LINK_TIME_OPTIMIZATION_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/snowkuma/rules.mk:4]:
```suggestion
LTO_ENABLE = yes
```

File: keyboards/crkbd/keymaps/snowkuma/rules.mk [hunk 11]
```diff
@@ -0,0 +1,7 @@
...
+
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no    # Breathing sleep LED during USB suspend
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/snowkuma/rules.mk:7]:
```suggestion
```

File: users/snowkuma/keycodes.h [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+#define INPUT_L LCAG(KC_SPC)
+#define TXT_PLS G(KC_PLUS)
+#define TXT_MIN G(KC_MINS)
+#define CUT G(KC_X)
+#define COPY G(KC_C)
+#define PASTE G(KC_V)
+#define NXT_WIN G(KC_GRV)
+#define EMOJI G(C(KC_SPC))  // brings up the emoji picker in OSX
+#define VOLUP KC__VOLUP     // shorter naming for layout tidiness
+#define VOLDOWN KC__VOLDOWN
```
Comment (fauxpark) [inline users/snowkuma/keycodes.h:68]: You should use `KC_VOLU` and `KC_VOLD` here instead, as they are recognised on all OSes.

File: users/snowkuma/rgblight_breathe_table.h [hunk 13]
```diff
@@ -0,0 +1,118 @@
+#ifndef RGBLIGHT_EFFECT_BREATHE_TABLE
+#define RGBLIGHT_EFFECT_BREATHE_TABLE
+
```
Comment (fauxpark) [inline users/snowkuma/rgblight_breathe_table.h:3]:
```suggestion
#pragma once

```

File: users/snowkuma/rgblight_breathe_table.h [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+  0x67,  0x60,
+  0x5a,  0x54,
+  0x4e,  0x48
+ #endif /* 64 bytes table */
+};
+// clang-format on
+
+static const int table_scale = 256 / sizeof(rgblight_effect_breathe_table);
+
+#endif /* RGBLIGHT_EFFECT_BREATHE_TABLE */
```
Comment (fauxpark) [inline users/snowkuma/rgblight_breathe_table.h:118]:
```suggestion
```

File: keyboards/planck/keymaps/snowkuma/keymap.c [hunk 15]
```diff
@@ -22,352 +22,104 @@
...
+		if (muse_counter == 0) {
+			uint8_t muse_note = muse_offset + SCALE[muse_clock_pulse()];
+			if (muse_note != last_muse_note) {
+				stop_note(compute_freq_for_midi_note(last_muse_note));
+				play_note(compute_freq_for_midi_note(muse_note), 0xF);
+				last_muse_note = muse_note;
+			}
+		}
+		muse_counter = (muse_counter + 1) % muse_tempo;
+	}
```
Comment (fauxpark) [inline keyboards/planck/keymaps/snowkuma/keymap.c:119]:
```suggestion
    if (muse_mode) {
        if (muse_counter == 0) {
            uint8_t muse_note = muse_offset + SCALE[muse_clock_pulse()];
            if (muse_note != last_muse_note) {
                stop_note(compute_freq_for_midi_note(last_muse_note));
                play_note(compute_freq_for_midi_note(muse_note), 0xF);
                last_muse_note = muse_note;
            }
        }
        muse_counter = (muse_counter + 1) % muse_tempo;
    }
```

File: keyboards/planck/keymaps/snowkuma/keymap.c [hunk 16]
```diff
@@ -22,352 +22,104 @@
...
+#endif
 }
 
 void matrix_init_user(void) {
-    set_unicode_input_mode(UC_OSX);
+	set_unicode_input_mode(UC_OSX);
```
Comment (fauxpark) [inline keyboards/planck/keymaps/snowkuma/keymap.c:124]:
```suggestion
    set_unicode_input_mode(UC_OSX);
```

**Pull request closed without merging**